tuning.py

```diff
@@ -0,0 +1,317 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Landscape tuning."""
+
+import itertools
+from pprint import pprint
+from typing import Optional, Tuple, Dict
+
+import pandas as pd
+import numpy as np
+from scipy.special import comb
+
+import potts_model
+import sampling
+import experiment
+import utils
+
+
+def get_fraction_adaptive_singles(landscape: potts_model.PottsModel) -> float:
+    """Returns the fraction of singles that are adaptive."""
+    all_singles = sampling.get_all_single_mutants(landscape.wildtype_sequence, landscape.vocab_size)
+    wt_fitness = landscape.evaluate(landscape.wildtype_sequence).item()
+    fraction_adaptive = (landscape.evaluate(all_singles) >= wt_fitness).sum() / all_singles.shape[0]
+    return fraction_adaptive
+
+
+def get_doubles_df(landscape: potts_model.PottsModel, threshold: float, adaptive: bool) -> pd.DataFrame:
+    """Returns a DataFrame of all pairwise combinations of singles above/below a given fitness threshold.
+
+    For each double mutant, the DataFrame includes the keys:
+        'fitness': the fitness of the double.
+        'a_fitness': the fitness of the first constituent single.
+        'b_fitness': the fitness of the second constituent single.
+        'residual': the epistatic term ].
+
+    Args:
+        landscape: The landscape on which to compute double fitness.
+        threshold: The threshold fitness for filtering single mutants.
+        adaptive: A boolean, indicating, if True (False), to construct pairwise combinations from single mutants
+            that are above (below) the `threshold` fitness.
+
+    Returns:
+        A DataFrame of double mutants with keys 'fitness', 'a_fitness', 'b_fitness', and 'residual'.
+    """
+    wt_seq = landscape.wildtype_sequence
+    all_singles = sampling.get_all_single_mutants(wt_seq, landscape.vocab_size)
+    df = experiment.get_fitness_df(all_singles, landscape.evaluate, wt_seq)
+    if adaptive:
+        singles = df[df.fitness >= threshold].sequence
+    else:
+        singles = df[df.fitness < threshold].sequence
+
+    if len(singles) < 2:
+        raise ValueError(f'Invalid Landscape: fewer than 2 singles for threshold {threshold}')
+
+    doubles = []
+    single_a = []
+    single_b = []
+    for a, b in itertools.combinations(singles, 2):
+        # check that this is a true double
+        if utils.get_mutation_positions(a, wt_seq) == utils.get_mutation_positions(b, wt_seq):
+            continue
+        double = utils.add_seqs(a, b, wt_seq)[0]
+        doubles.append(double)
```

Comment:
style tip: when building up the entries of a dataframe, I find it cleaner if you build a list of dicts, where each dict represents a record, like 
records = []
for i in range(10):
  records.append(dict(a=i, b=i+1))
df = pd.DataFrame(records)